File: BenefitCostPreview/Logic/BenefitCostService.cs
using BenefitCostPreview.Interfaces;
using BenefitCostPreview.Models;
using System;
using System.Collections.Generic;
using System.Linq;

namespace BenefitCostPreview.Logic
{
    public class BenefitCostService : IBenefitCostService
    {
        private const decimal EmployeeYearlyBenefitCost = 1000;
        private const decimal DependentYearlyBenefitCost = 500;
        private const decimal NumPaychecks = 26;
        private const decimal EmployeePaycheck = 2000;

        public PaycheckSummary CalculateBenefitCostPreview(string employeeFirstName, IEnumerable<string> dependentFirstNames = null)
        {
            dependentFirstNames = dependentFirstNames?.ToList();

            if (string.IsNullOrEmpty(employeeFirstName) ||  dependentFirstNames != null && dependentFirstNames.Any(string.IsNullOrEmpty))
                throw new Exception("All names must have a value");

            var yearlyBenefitCost = CalculateEmployeeYearlyCost(employeeFirstName);
            if (dependentFirstNames?.Any() ?? false)
            {
                yearlyBenefitCost += dependentFirstNames
                    .Select(CalculateDependentYearlyCost)
                    .Sum();
            }

            var benefitsCost = yearlyBenefitCost / NumPaychecks;

            return new PaycheckSummary{
                BenefitsCost = benefitsCost,
                TakeHomePay = EmployeePaycheck - benefitsCost
            };
        }

        private decimal CalculateEmployeeYearlyCost(string name)
        {
            var discountFactor = GetDiscountFactor(name);
            return CalculateCost(EmployeeYearlyBenefitCost, discountFactor);
        }

        private decimal CalculateDependentYearlyCost(string name)
        {
            var discountFactor = GetDiscountFactor(name);
            return CalculateCost(DependentYearlyBenefitCost, discountFactor);
        }

        private decimal CalculateCost(decimal baseCost, decimal discountFactor)
        {
            return baseCost * discountFactor;
        }

        private decimal GetDiscountFactor(string name)
        {
            return name[0] switch
            {
                'A' => 0.9M,
                'a' => 0.9M,
                _ => 1
            };
        }
    }
}

File: BenefitCostPreview/ClientApp/src/app/services/benefit-cost/benefit-cost.service.ts
import { Injectable } from '@angular/core';
import { PaycheckSummary } from 'src/app/models/paycheck-summary';


@Injectable({
  providedIn: 'root'
})
export class BenefitCostService {

  private readonly employeeYearlyBenefitCost = 1000;
  private readonly dependentYearlyBenefitCost = 500;
  readonly numPaychecks = 26;
  readonly employeePaycheck = 2000;

  constructor() { }

  calculatePaycheckPreview(employeeFirstName: string, dependentFirstNames: string[] = []): PaycheckSummary {
    if (!employeeFirstName || (dependentFirstNames && dependentFirstNames.some(d => !d))) {
      throw Error('All names must have a value');
    }

    let yearlyBenefitCost = this.calculateEmployeeYearlyCost(employeeFirstName);
    if (dependentFirstNames) {
      yearlyBenefitCost += dependentFirstNames
        .map(d => this.calculateDependentYearlyCost(d))
        .reduce((a, b) => a + b, 0);
    }

    const benefitsCost = yearlyBenefitCost / this.numPaychecks;

    return {
      benefitsCost,
      takeHomePay: this.employeePaycheck - benefitsCost
    };
  }

  private calculateEmployeeYearlyCost(name: string): number {
    const discountFactor = this.getDiscountFactor(name);
    return this.calculateCost(this.employeeYearlyBenefitCost, discountFactor);
  }

  private calculateDependentYearlyCost(name: string): number {
    const discountFactor = this.getDiscountFactor(name);
    return this.calculateCost(this.dependentYearlyBenefitCost, discountFactor);
  }

  private calculateCost(baseCost: number, discountFactor: number): number {
    return baseCost * discountFactor;
  }

  private getDiscountFactor(name: string): number {
    switch (name[0]) {
      case 'A':
      case 'a':
        return 0.9;
      default:
        return 1;
    }
  }
}

